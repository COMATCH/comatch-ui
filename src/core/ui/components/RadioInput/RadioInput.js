import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import noop from 'lodash/noop';
import { InputLabel } from '../InputLabel';
import { InputError } from '../InputError';

import './RadioInput.scss';

const propTypes = {
    checked: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
    disabled: PropTypes.bool,
    display: PropTypes.oneOf(['block', 'inline']),
    id: PropTypes.string,
    inputError: PropTypes.string,
    label: PropTypes.string,
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
};

const defaultProps = {
    checked: null,
    disabled: false,
    display: 'block',
    id: null,
    inputError: '',
    label: '',
    onChange: noop,
    value: '',
};

export const RadioInput = ({ checked, disabled, display, id, inputError, label, onChange, name, value }) => {
    const classes = classNames('RadioInput', display, { disabled, 'has-error': inputError });

    return (
        <div id={id} className={classes}>
            <label>
                <input
                    type="radio"
                    id={id}
                    name={name}
                    onChange={onChange}
                    value={value}
                    checked={checked}
                    disabled={disabled}
                />
                <div className="RadioInput__input" role="radio" />
                {label && <InputLabel className="RadioInput__label" span={true} text={label} />}
            </label>
            {inputError && <InputError text={inputError} />}
        </div>
    );
};

RadioInput.defaultProps = defaultProps;

RadioInput.propTypes = propTypes;
