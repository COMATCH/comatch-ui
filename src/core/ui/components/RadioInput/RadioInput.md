
## RadioInput

From [`RadioInput.js`](RadioInput.js)



prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**checked** | `Union<Boolean \| String>` | `null` | :x: | 
**disabled** | `Boolean` | `false` | :x: | 
**display** | `Enum('block','inline')` | `'block'` | :x: | 
**id** | `String` | `null` | :x: | 
**inputError** | `String` | `''` | :x: | 
**label** | `String` | `''` | :x: | 
**name** | `String` |  | :white_check_mark: | 
**onChange** | `Function` | `noop` | :x: | 
**value** | `Union<String \| Number \| Boolean>` | `''` | :x: | 



