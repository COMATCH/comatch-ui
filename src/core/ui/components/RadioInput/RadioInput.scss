@import '../../../styles/variables.scss';

// Radio inputs
.RadioInput {
    display: inline-block;
    text-align: center;
    position: relative;

    // Displays
    &.block {
        display: block;
        text-align: left;
    }

    // Whitespace
    &:not(.block) ~ &:not(.block) {
        margin-left: 30px;
    }

    &.block ~ &.block {
        margin-top: 5px;
    }

    // Elements
    &__input {
        background-color: $co-color-white;
        border: 1px solid $co-color-mid-gray;
        border-radius: 50%;
        width: 16px;
        height: 16px;
        position: relative;

        &:after {
            background-color: $co-color-primary;
            border-radius: 50%;
            display: block;
            content: '';
            position: absolute;
            top: 50%;
            left: 50%;
            width: 8px;
            height: 8px;
            transition: transform 400ms cubic-bezier(0.45, 1.8, 0.5, 0.75);
            transform: scale(0) translate3d(-50%, -50%, 0);
            transform-origin: top left;
        }

        input[type="radio"]:disabled ~ &:after {
            background-color: $co-color-mid-gray;
        }

        // Whitepsace
        & ~ .InputLabel {
            margin-left: 10px;
        }
    }

    &__input,
    .InputLabel {
        display: inline-block;
        vertical-align: middle;
        color: $co-color-black;
        font-size: $co-font-sm;
        font-weight: normal;
    }

    input[type="radio"]:disabled ~ &__label {
        color: $co-color-mid-gray;
    }

    label {
        cursor: pointer;
        margin: 0;
    }

    input[type='radio'] {
        display: none;

        &:checked ~ .RadioInput__input:after {
            transform: scale(1) translate3d(-50%, -50%, 0);
        }
    }

    span.InputLabel {
        margin-bottom: 2px;
    }

    // States
    &.disabled {
        * {
            color: $co-color-black !important; // important can be removed when legacy css is removed
        }

        label {
            cursor: not-allowed;
        }
    }
}
