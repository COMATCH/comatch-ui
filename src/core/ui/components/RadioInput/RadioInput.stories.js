import React from 'react';
import { storiesOf } from '@storybook/react';
import { RadioInput } from './RadioInput';

storiesOf('RadioInput', module)
    .add('unchecked with label', () => <RadioInput name="fruit" label="Label" />)
    .add('checked with label', () => <RadioInput name="fruit" label="Label" checked />);
