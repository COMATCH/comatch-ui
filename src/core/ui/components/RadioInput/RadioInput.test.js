import React from 'react';
import renderer from 'react-test-renderer';
import '../../../config/tests/setup';
import { mount } from 'enzyme';
import { lorem } from 'faker';
import noop from 'lodash/noop';
import { RadioInput } from './RadioInput';

describe('RadioInput tests', () => {
    const REQUIRED_PROPS = {
        name: 'some name',
    };

    const OPTIONAL_PROPS = {
        className: 'RadioInput block',
        checked: null,
        disabled: false,
        id: lorem.word(),
        name: lorem.word(),
        onChange: jest.fn(),
        type: 'radio',
        value: lorem.word(),
        label: 'Label text'
    };

    it('should create and compare snapshot', () => {
        const radioInput = renderer.create(<RadioInput {...REQUIRED_PROPS} />).toJSON();
        expect(radioInput).toMatchSnapshot();
    });

    it('should render correctly with only required props', () => {
        const props = {
            ...REQUIRED_PROPS,
        };
        const radioInputComponent = mount(<RadioInput {...props} />);
        expect(radioInputComponent.prop('checked')).toEqual(null);
        expect(radioInputComponent.prop('disabled')).toEqual(false);
        expect(radioInputComponent.prop('display')).toEqual('block');
        expect(radioInputComponent.prop('id')).toEqual(null);
        expect(radioInputComponent.prop('inputError')).toEqual('');
        expect(radioInputComponent.prop('label')).toEqual('');
        expect(radioInputComponent.prop('onChange')).toEqual(noop);
        expect(radioInputComponent.prop('name')).toEqual('some name');
        expect(radioInputComponent.prop('value')).toEqual('');
    });

    it('should render correctly with optional props', () => {
        const props = {
            ...REQUIRED_PROPS,
            ...OPTIONAL_PROPS,
        };
        const radioInputComponent = mount(<RadioInput {...props} />);
        expect(radioInputComponent.prop('className')).toEqual(props.className);
        expect(radioInputComponent.prop('checked')).toEqual(props.checked);
        expect(radioInputComponent.prop('disabled')).toEqual(props.disabled);
        expect(radioInputComponent.prop('id')).toEqual(props.id);
        expect(radioInputComponent.prop('name')).toEqual(props.name);
        expect(radioInputComponent.prop('type')).toEqual(props.type);
        expect(radioInputComponent.prop('value')).toEqual(props.value);
    });

    describe('verifying the HTML structure', () => {
        const props = {
            ...REQUIRED_PROPS,
            ...OPTIONAL_PROPS,
        };
        it('should expect nodes/dom elements', () => {
            const radioInputComponent = mount(<RadioInput {...props} checked={true} />);
            expect(radioInputComponent).toHaveLength(1);
            expect(radioInputComponent.find('div')).toHaveLength(2);

            expect(radioInputComponent.find('div').find('label')).toHaveLength(1);

            expect(radioInputComponent.find('div').find('label').find('input')).toHaveLength(1);
            expect(radioInputComponent.find('div').find('label').find('input').at(0).prop('type')).toBe(props.type);
            expect(radioInputComponent.find('div').find('label').find('input').at(0).prop('checked')).toBe(true);

            expect(radioInputComponent.find('div').find('label').find('div').hasClass('RadioInput__input')).toBe(true);

            expect(radioInputComponent.find('div').find('span')).toHaveLength(1);
            expect(radioInputComponent.find('div').find('span').text()).toBe(props.label);
            expect(radioInputComponent.find('div').find('span').hasClass('InputLabel RadioInput__label')).toBe(true);
        });
    });
});
