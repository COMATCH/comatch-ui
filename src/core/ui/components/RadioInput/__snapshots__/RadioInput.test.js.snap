// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioInput tests should create and compare snapshot 1`] = `
<div
  className="RadioInput block"
  id={null}
>
  <label>
    <input
      checked={null}
      disabled={false}
      id={null}
      name="some name"
      onChange={[Function]}
      type="radio"
      value=""
    />
    <div
      className="RadioInput__input"
      role="radio"
    />
    
  </label>
  
</div>
`;
